src/wasm.rs

```diff
@@ -163,12 +174,29 @@ where
                 let addr = api.addr_validate(&contract_addr)?;
                 let contract = self.load_contract(storage, &addr)?;
                 let mut res = ContractInfoResponse::default();
-                res.code_id = contract.code_id as u64;
+                res.code_id = contract.code_id;
                 res.creator = contract.creator.to_string();
                 res.admin = contract.admin.map(|x| x.into());
                 to_binary(&res).map_err(Into::into)
             }
-            query => bail!(Error::UnsupportedWasmQuery(query)),
+            #[cfg(feature = "cosmwasm_1_2")]
+            WasmQuery::CodeInfo { code_id } => {
+                let Some(code_data) = self.codes.get(&code_id) else {
+                    bail!(Error::UnregisteredCodeId(code_id))
+                };
+                let mut res = cosmwasm_std::CodeInfoResponse::default();
+                res.code_id = code_id;
+                res.creator = code_data.creator.to_string();
+                res.checksum = cosmwasm_std::HexBinary::from(
+                    Sha256::digest(format!(
+                        "contract code {} stored by {}",
```

Comment:
I like the idea that the code-id shouldn't impact the contract checkseum. Maybe we should add a name method to the Contract trait ? And that could be used to derive a checksum sort of string . But that's an additional unnecessary parameter that projects have to implement. 
However, I think that whatever the function is, it should be a separate function that is called here (for tests or even for using the library, the function checksum function should be accessible outside of the crate)

Reply:
This case is little bit tricky. In real blockchain, we deploy a blob and want to get a hash of it. Every change of the contract's code will produce a different hash code. But do we care about the value of this hash? I guess not. It doesn't matter what value it has. It does matter that this is the same value all times the same contract is deployed. The same is during testing. We want to get the same value of the hash each time the test is run. But unlike in real blockchain, we don't want the hash to be changed each time the contract's code changes, because the we would have to change the test before each run - this is a waste of time. In test examples I have seen, maximum 2 contracts are deployed and tested. Usually the same creator could be used. In such case the `code_id` is the differentiator for the hash. And the hash is the same each time the test is run. Let me know what you think.

Reply:
> Maybe we should add a name method to the Contract trait
The problem is that now I can create two completely separate contracts with the same name and checksums. It is an invalid and error-prone approach (the idea is that I should be able to depend on the fact that checksum is the same = behavior is the same). Sure, there are SHA collisions, but we can agree on this level of sacrifice.
This is the test framework, so it is not a big deal. In most cases, you will not care about hash, as it is a rare thing to use. And you will name your contracts the same for the sake of simplicity. But at some point, you will care. You are aware, dev. You will know which contract should have a different name. Then, at some point, some less aware dev would use your contract in his testing environment, and he will not even know it uses hash internally (it is black-box to him), and he would not maybe even understand how those hashes work. And he will make the same name for multiple contracts, and with some screwed-up logic, your contract misbehaves.

The problem doesn't exist if you hash the wasm binary code, as the wasm binary code = it works. The only thing I can imagine is to try to build the passed `dyn Contract` as a WASM binary to hash it, but it is super tricky and time-consuming. I don't think there's any point in that.

The way, I think, may be to have some API to duplicate the already stored code. Some API like:
```rust
fn duplicate_code(&mut self, code_id: u64) -> u64;
```
This would create a new code id using the same code_id as passed. Because the `dyn Contract` is not passed again, MT can guarantee the same behavior of both contracts. By doing so, we can evaluate both those code IDs to the exact same hash. It won't work if we evaluate hash from the id itself, but we can add an additional field - `seed: u64` which would be the seed to calculate hash. By default (with `store_code`), we would make the seed to be the returned `code_id`, but when we duplicate code, we can copy the `seed` from the previous code. I personally like this approach.

Note that at the end of the day - different hashes for the same behavior are the proper behavior. You can implement the same behavior differently or compile contracts with different cargo versions - the binary might slightly differ and it is all fine. Change a single constant in your binary (contract version, for eg), and you are changing hash. So technically, having a different hash for the same contracts in this case is not the end of the world, but it would be nice to simulate the situation where multiple code ids have actually the same hash, which is a case possible in the real world.

Reply:
So, we leave it as it is?

Reply:
For now we can leave it, but I would add the `duplicate_code` as well (possibly in a separate PR).

Reply:
So, finally I remove creator from hash generation and leave only `code_id`.